### PR TITLE
MGMT-1844 Improve errors handling for Auth and Autz

### PR DESCRIPTION
**Pull request by rollandf**
- Define infra_error model in swagger.yaml
- Update all 401 and 403 to use infra_error
- Update inventory.go that created a Forbidden error to use infra_error (https://github.com/openshift/assisted-service/blob/master/internal/bminventory/inventory.go#L1172)
- Update Authenticator to return `false` if Auth did not suceed

With these changes, all errors from Auth will be of type 401 and there will be no Unmarshall Error on clients side.
Agent and Assisted-Installer will need to consume the new API after this PR is merged.

### CODE REVIEW

**Comment by rollandf** [comment]
@ronniel1 @eranco74 @ori-amizur @avishayt @nmagnezi @danielerez

**Comment by avishayt** [comment]
@rollandf Why can't we use the existing error, setting 'code' and 'reason'?  That is the error format that app-sre requested.

**Comment by rollandf** [comment]
> @rollandf Why can't we use the existing error, setting 'code' and 'reason'? That is the error format that app-sre requested.

These 401/403 errors are created by swagger infra and not by assisted-service. We don't have control on how they are created.

**Comment by rollandf** [comment]
This the error we get in Agent currently:

`Aug 20 07:37:07 test-infra-cluster-master-2 systemd[1]: Started agent.service.
Aug 20 07:37:09 test-infra-cluster-master-2 agent[1745]: time="20-08-2020 07:37:09" level=warning msg="Error registering host: json: cannot unmarshal number into Go struct >`

After the fix:

`Aug 20 13:18:47 test-infra-cluster-master-2 agent[1828]: time="20-08-2020 13:18:47" level=warning msg="Error registering host: [POST /clusters/{cluster_id}/hosts][401] registerHostUnauthorized  &{Code:0xc0003a4148 Message:0xc0004401e0}" file="register_node.go:47" request_id=cb37c18f-b977-49da-856b-391ca116b889`

File: swagger.yaml [hunk 1]
```diff
@@ -2205,6 +2205,22 @@ definitions:
...
+    type: object
+    required:
+      - code
+      - message
+    properties:
+      code:
+        type: integer
+        format: int32
+        description: Numeric identifier of the error.
+        minimum: 400
```
Comment (avishayt) [inline swagger.yaml:2218]: 401

File: swagger.yaml [hunk 2]
```diff
@@ -2205,6 +2205,22 @@ definitions:
...
+        maximum: 504
```
Comment (avishayt) [inline swagger.yaml:2219]: 403

**Review by avishayt — COMMENTED** [review]
I don't love this solution.  It's very confusing to users to deal with two different error structs.  There's absolutely no way to use the existing one?

**Comment by rollandf** [comment]
@ronniel1 PTAL

**Comment by rollandf** [comment]
@ronniel1 ping

**Comment by avishayt** [comment]
@ronniel1 Do you know any way that the swagger infra can return different errors?

**Comment by ronniel1** [comment]
> I don't love this solution. It's very confusing to users to deal with two different error structs. There's absolutely no way to use the existing one?

@avishayt for now we don't have a way.
We have Jira ticket on it: https://issues.redhat.com/browse/MGMT-1097

**Comment by avishayt** [comment]
@ronniel1 OK, so it's fine for me if it's fine for you.

File: swagger.yaml (hunk 2 above)
Comment (rollandf) [reply swagger.yaml]: Done

File: swagger.yaml (hunk 1 above)
Comment (rollandf) [reply swagger.yaml]: Done

**Review by avishayt — APPROVED** [review]

**Pull request merged**